### ISSUE TITLE
Implement Nix JSON RPC interface

### ISSUE THREAD

**Issue by gytis-ivaskevicius** [NixOS/nix#8726]
**Is your feature request related to a problem? Please describe.**
I would like for it to be easier to integrate with nix

**Describe the solution you'd like**
Nix JSON RPC (or similar) interface

**Describe alternatives you've considered**
Currently, I am executing nix commands, in some cases I am forced to parse text (like to figure out if the derivation is uncached I do `nix-store --realise --try-run <drv>`)

**Additional context**
Expected features:
- More or less whole nix CLI functionality to be to be available via JSON
- Every  JSON request should result in async event stream of which files have been evaluated, which derivations have been built during eval/build time, which derivations have been fetched
- Not necessary but it would be great to have daemon mode to reuse eval cache and nix binary querying cache (like which derivations have not been found)

**Future work**
I think It would be a good idea to separate nix and CLI. And if we ever have that - it would be cool to have Nix CLI download Nix based on whatever is pinned in `flake.lock` thus further reducing impurities

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by thufschmitt** [comment]
This should be implementable out-of-tree once #8699 lands
